Add silent-payments-output-descriptors review to bitcoinplusplus/sovereignty-edition (#700)
bitcointranscripts/bitcointranscripts: bitcoinplusplus/sovereignty-edition/silent-payments-output-descriptors.md

Reviewer: ```suggestion
Because the idea with a silent payment address is that you could post it on your website, but you could have one on your website, maybe one on your GitHub, another one on your Twitter profile, and using the labels you could have give each of them a different label.
```

Reviewer: ```diff
- Because they deal with
+ Because the idea with ...
```
<img width="2610" height="1204" alt="Image" src="https://github.com/user-attachments/assets/992ded8c-d03b-420f-88c7-b3647a16791d" />

---
title: 'Silent Payments: Output Descriptors'
transcript_by: 'muchai254 via review.btctranscripts.com'
media: 'https://youtu.be/QQO0pMQB-QU'

> date: '2026-01-17'

Reviewer: ```suggestion
date: '2025-12-16'
```

Reviewer: <img width="2067" height="1088" alt="Image" src="https://github.com/user-attachments/assets/e58f060c-dd18-4621-b905-158d1fe883f1" />

---
## Introduction to Silent Payment Descriptors 

niftynei: 00:00:00

Great.
Okay, so silent payments are pretty cool.
We're not going to talk about that.
In fact, there's already a BIP, BIP352 for silent payments.
What this talk is about is about the add output descriptor format.
Specifically, the BIP proposal for adding an output descriptor format.
Okay, so this is the BIP.

> It's called Add sp() Output Descriptor Format for BIP 352, which is the BIP number for silent payments.

Reviewer: ```suggestion
It's called Add `sp()` Output Descriptor Format for BIP 352, which is the BIP number for silent payments.
```

How many of you have heard about this proposal for output descriptors.
This is new, right?
This is new for most people.
Okay, cool.
Great.
So, okay.
So, the specification is pretty short.

> It says we're going to add a new top level script expression as defined as sp().

Reviewer: ```suggestion
It says we're going to add a new top level script expression as defined as `sp()`.
```

And then it also adds two new things called key expressions.
We're going to learn about these in a second.
They're new key expression types.
One is called `spscan` and the other is called `spspend`.
So we're going to have two new ways of expressing a particular key.
The difference between them is that one has a private key encoded in it and a public key encoded in it.
The other one, `spspend`, has two private keys encoded in it.
And I'll show you what this looks like in a second.
So don't worry if this isn't exactly understandable.
Okay, so you're like, okay, what's the difference between `spscan` and `spspend`?
`spscan` is for watch-only wallets, because it has a public key part of it in it.
And then `spspend` as a key encoding is going to have full wallets because it has two private keys.
So basically if someone sends you, 
So the key idea with a key expression is that it's like, okay, here is key material.
This will either let you identify coins that have been locked up to this particular information, or it will let you both identify them and also spend them.
But I'll get into this a little more explicitly.

> ## What sp() Descriptors Represent

Reviewer: ```suggestion
## What `sp()` Descriptors Represent
```

niftynei: 00:02:03

Okay, so let's just go back to like, okay, what are we talking about?

> We're talking about this sp() thing.

Reviewer: ```suggestion
We're talking about this `sp()` thing.
```

So SP is short for silent payments.
So basically when you have a descriptor, what a descriptor, what the whole point of having an output descriptor is that it's like here's a description of what an outpoint in Bitcoin might look like on chain and it's a way that if you have this description of what an outpoint or output might look like you're then able to go and find them across all of the outpoints that exist in Bitcoin currently.
Both past ones that have been spent and ones that are eligible to be spent, to spend.
And the whole goal of this is that your wallet will be able to identify coins that you're able, interested, they're either interested in keeping track of when they're spent versus not, or, and maybe additionally, able to spend those outputs, if that makes sense.
So we basically need a way of writing down, these are outputs that I'm interested in, and here's information that you would need in order to be able to spend them in some cases.
OK, so we're going to write this as kind of like it's very, for whatever reason, all output descriptors look like function definitions.

> So this one is sp(), so it's like, OK, for a silent payment, here's some information.

Reviewer: ```suggestion
So this one is `sp()`, so it's like, OK, for a silent payment, here's some information.
```

So the first thing that we're going to put inside of this kind of set of information about a silent payment output is something called a key expression.

## Why Silent Payments Need New Key Expressions

niftynei: 00:03:23

How many of you have seen XPUBs before?
So, XPUB is kind of a traditional or classic key expression.
It gives you information to be able to find not only one output, but a series of outputs that have been spent to that particular output descriptor, to that particular XPUB, right?
XPUBs have this property where you can spend multiple ones of them, and they have a chain code included in it, etc., etc.

> XPUBs are not valid key expressions for this sp() descriptor.

Reviewer: ```suggestion
XPUBs are not valid key expressions for this `sp()` descriptor.
```

Instead, Craig invents two new ways of encoding information about keys.

> In order to put a key expression inside of this sp() thing, and this is part of the spec, which I could pull up to show you, but that's okay.

Reviewer: ```suggestion
In order to put a key expression inside of this `sp()` thing, and this is part of the spec, which I could pull up to show you, but that's okay.
```

So it must be a silent payment key expression, which he defines two, which we kind of just looked at.
One is this `spscan`, and the other is `spspend`.
So these are actually gonna be long series of information, a lot like an XPUB is.
XPUBs can be quite long.
They're encoded in Bech32.

> (XPUBs are encoded using Base58Check. The new `spscan` and `spspend` key expressions are encoded in Bech32)

Reviewer: Nicely done! I really like this clarification note

> (XPUBs are encoded using Base58Check. The new `spscan` and `spspend` key expressions are encoded in Bech32)

Reviewer: ```suggestion
(XPUBs are encoded using Base58Check. The new `spscan` and `spspend` key expressions are encoded in Bech32m)
```

Reviewer: > The spscan key expression encodes the scan private key and spend public key. It is a [Bech32m](https://bips.dev/350) encoding of:
> 
> The spspend key expression encodes both the scan and spend private keys. It is a [Bech32m](https://bips.dev/350) encoding of:

https://bips.dev/392/

> After a [problem](https://bitcoinops.org/en/newsletters/2019/11/13/#taproot-review-discussion-and-related-information) was discovered with bech32 error detection for future upgrades under some rare circumstances, a new bech32 modified (bech32m) format was proposed

https://bitcoinops.org/en/topics/bech32/

<img width="1524" height="438" alt="Image" src="https://github.com/user-attachments/assets/bc9861e3-5088-4f6f-b0ef-6f09c128d759" />

https://bitcoinops.org/en/newsletters/2026/03/13/

I don't really care about the encoding.
All I care about is what information is inside of them.
So what exactly goes in this dot, dot, dot portion of a key expression for a silent payment?
For, okay, so this is from the BIP proposal that Craig wrote.
He's got `spscan` and `spspend`.
So actually, he kind of then defines the data part.
It starts with the character "q", because that's V0.
And then there's a payload in each of them.
I'm not going to look at the actual BIP, but I'm just going to kind of explain what's in them.

## The `spscan` Key Expression

niftynei: 00:05:03

> So for `spscan`, the first kind of piece of information that you're going to have in this Bech32 string is a private key.

Reviewer: ```suggestion
So for `spscan`, the first kind of piece of information that you're going to have in this Bech32m string is a private key.
```

Reviewer: I think she misspoke a bit. Should we just fix it, rather than leaving a note like `Bech32(Bech32m)` ? What do you think?

<img width="2166" height="952" alt="Image" src="https://github.com/user-attachments/assets/a6d7c888-b0e4-475e-b983-d46375117110" />

All private key material in Bitcoin is 32 bytes of random data, so it's just a number, and this is no different, so you're going to have a 32 byte piece of information, which is a private key, and that's gonna be your scan key, and then there's gonna be a second key inside of this basically key expression, and for the scan type of it, it's actually gonna be a public key.
This is gonna be our spend key.
So spend key basically is whether or not you're able to move the funds that are locked up to that outpoint.
We just have the public key and the `spscan` type of key description.
So this is public key here.
It's gonna be 33 bytes because if you're familiar with how public keys in elliptic curve encryption work, you'll know that they're actually points.
It's x, y, and we put them in the compressed form.
So it's always 33 bytes long.
That's because it's an X coordinate plus a parity description for which Y you pick.
If you're not familiar with private and public keys, there's a great Base58 class all about cryptography up on Udemy.
So I encourage you to go check that out where we explain exactly how private and public keys work for elliptic curves.

> But yeah, so for `spscan`, you're gonna have, again, two kind of pieces of information encoded into a Bech32 description.

Reviewer: ```suggestion
But yeah, so for `spscan`, you're gonna have, again, two kind of pieces of information encoded into a Bech32m description.
```

Hopefully that happens soon.
Yeah, I also run bitcoin++.
Thanks for coming to our first ever sovereignty edition in Taipei.
We have about eight more minutes.
If anyone has any questions, happy to hear them.

## Q&A

[Audience]: 00:18:08

Thanks nifty.
Might be out of scope, but do you know if there's any plan to be able to derive the scan private key from the spend private key, like deterministically?

niftynei: 00:18:19

Yeah, that's a great question.
I think they have to stay separate.
And I can pull up, I mean, if you want to be, let me see if I can find the...
I don't have a link to the BIP, do I?
That would be handy.
BIP 352, there it is, that's this one.
Is there a way to full screen this?
There's gotta be a way to full screen this.
There we go, does that work?
Ta-da, kind of.
Okay, so.
Okay, so this is the silent payments BIP, which I have not completely read.
I shouldn't admit that on camera.
It's OK.
So if you dig into the math of how these work, you want them to be separate.
So you could use them.
They could be the same thing.
And in fact, the original proposal, if you look at the simple case, this integer A, I think, discovers, no, that's not it.

> Okay, so in the original, kind of the simplest case of single-key silent payments.

Reviewer: She said "private payments". Is she wrong? Did she misspeak?

<img width="2079" height="693" alt="Image" src="https://github.com/user-attachments/assets/939dfe88-e6f5-4f93-a9a9-edb44a909dd3" />

> Okay, so in the original, kind of the simplest case of single-key silent payments.

Reviewer: ```suggestion
Okay, so in the original, kind of the simplest case of private payments.
```

Reviewer: https://github.com/bitcoin/bips/blob/master/bip-0352.mediawiki

<img width="1675" height="160" alt="Image" src="https://github.com/user-attachments/assets/24b804af-2f81-4f3e-b527-b3c33bd839e8" />

Sorry, does that answer your question?
That was a longer.

[Audience]: 00:23:14

Do you have a backup path?
So you can use one backup for both the scan and spend keys?

niftynei: 00:23:27

Ideally, yeah, probably.
But you wouldn't want to, if you have the spend key and you're deriving the scan key, yeah, this is beyond my pay grade at this point.
But yeah, you're right.
If you had a way of deriving the scan key from the spend key, then you wouldn't need to encode both of them in the descriptor, right?
Because there would be, you would need to communicate that, it would be implicit.

> So maybe that's something to contribute back to the descriptors back, yeah.

Reviewer: ```suggestion
So maybe that's something to contribute back to the descriptors spec, yeah.
```

Reviewer: <img width="1613" height="373" alt="Image" src="https://github.com/user-attachments/assets/13c6066b-21ad-4abe-a117-5f02dab2cb95" />

[Audience]: 00:23:58

So I just have a comment.
You asked a question for Craig about optional birthday.
So since labels are numbers, probably it can't be.
There's no way to distinguish whether it's a label or a birthdate.

niftynei: 00:24:15

You're saying with making, so basically my question was like, could we make the birthdate optional and also include labels, right?
And I think you're right, because it's all the same type.
I don't know, something to think about.
I don't think leaving birthdates out is necessarily desirable.
But, cool.
Any other questions?
We've got like another minute and a half.

## Closing Remarks

niftynei: 00:24:39

This is the whole specification for silent payments.
Fun fact, silent payments require Schnorr, so they wouldn't be possible without Taproot.
For those of you who know much about the upgrade path on their cryptography stuff, we added Schnorr with Taproot.
I don't think you can't do the math that they're doing to make and have this all work prior to Taproot outputs, so silent payments must be Taproot outputs.
It's kind of cool.
Anything else?
Any other questions?
Okay.

> I will leave the floor a minute early.

Reviewer: ```suggestion
I will cede the floor a minute early.
```

Reviewer: <img width="1626" height="406" alt="Image" src="https://github.com/user-attachments/assets/df3c5e63-ac37-4757-b1f6-65f61d7e9296" />
